consensus/istanbul/backend/announce.go

```diff
@@ -70,6 +72,102 @@ const (
 	LowFreqState
 )
 
+// AddressProvider provides the different addresses the announce manager needs
+type AddressProvider interface {
+	SelfNode() *enode.Node
+	ValidatorAddress() common.Address
+	IsValidating() bool
+}
+
+type ProxyContext interface {
+	GetProxiedValidatorEngine() proxy.ProxiedValidatorEngine
+}
+
+type AnnounceSupport interface {
+	// Gossip gossips protocol messages
+	Gossip(payload []byte, ethMsgCode uint64) error
+	// RetrieveValidatorConnSet returns the validator connection set
+	RetrieveValidatorConnSet() (map[common.Address]bool, error)
+	// Multicast will send the eth message (with the message's payload and msgCode field set to the params
+	// payload and ethMsgCode respectively) to the nodes with the signing address in the destAddresses param.
+	Multicast(destAddresses []common.Address, payload []byte, ethMsgCode uint64, sendToSelf bool) error
+}
+
+type AnnounceManagerConfig struct {
+	IsProxiedValidator bool
+	AWallets           *atomic.Value
+	VcDbPath           string
+}
+
+type AnnounceManager struct {
+	logger log.Logger
+
+	config AnnounceManagerConfig
+
+	addrProvider AddressProvider
+	proxyContext ProxyContext
+	support      AnnounceSupport
+
+	valEnodeTable *enodes.ValidatorEnodeDB
+
+	versionCertificateTable *enodes.VersionCertificateDB
+
+	gossipCache GossipCache
+
+	// The enode certificate message map contains the most recently generated
+	// enode certificates for each external node ID (e.g. will have one entry per proxy
+	// for a proxied validator, or just one entry if it's a standalone validator).
+	// Each proxy will just have one entry for their own external node ID.
+	// Used for proving itself as a validator in the handshake for externally exposed nodes,
+	// or by saving latest generated certificate messages by proxied validators to send
+	// to their proxies.
+	enodeCertificateMsgMap     map[enode.ID]*istanbul.EnodeCertMsg
+	enodeCertificateMsgVersion uint
+	enodeCertificateMsgMapMu   sync.RWMutex // This protects both enodeCertificateMsgMap and enodeCertificateMsgVersion
+
+	lastVersionCertificatesGossiped   map[common.Address]time.Time
+	lastVersionCertificatesGossipedMu sync.RWMutex
+
+	lastQueryEnodeGossiped   map[common.Address]time.Time
+	lastQueryEnodeGossipedMu sync.RWMutex
+}
+
+// NewAnnounceManager creates a new AnnounceManager using the valEnodeTable given. It is
+// the responsibility of the caller to close the valEnodeTable, the AnnounceManager will
+// not do it.
+func NewAnnounceManager(config AnnounceManagerConfig, support AnnounceSupport, proxyContext ProxyContext,
+	addrProvider AddressProvider, valEnodeTable *enodes.ValidatorEnodeDB,
+	gossipCache GossipCache) *AnnounceManager {
+	am := &AnnounceManager{
+		logger:                          log.New(),
```

Comment:
you can add something like `log.New("module", "announceManager")`

Reply:
done